File: dotnet/Allmark/Parse/ParseIndent.cs
namespace Allmark.Parse;

using Allmark.Types;

public static class ParseIndent
{
    public static void Execute(BlockParserState state)
    {
        if (state.I < state.Src.Length && Utils.IsSpace(state.Src[state.I]))
        {
            var start = state.I;
            for (; state.I < state.Src.Length; state.I++)
            {
                char c = Utils.GetChar(state.Src, state.I);
                if (c == ' ')
                {
                    state.Indent += 1;
                }
                else if (c == '\t')
                {
                    state.Indent += 4 - (state.Indent % 4);
                }
                else if (c == '\n')
                {
                    state.HasBlankLine = true;
                    break;
                }
                else if (c == '\r')
                {
                    state.HasBlankLine = true;
                    if (Utils.GetChar(state.Src, state.I + 1) == '\n')
                    {
                        state.I++;
                    }
                    break;
                }
                else
                {
                    break;
                }
            }
            state.Spaces = state.Src.Substring(start, state.I - start);
        }
    }
}

File: zig/src/parse/parseIndent.zig
const std = @import("std");
const BlockParserState = @import("../types/BlockParserState.zig").BlockParserState;
const isSpace = @import("../utils/isSpace.zig").isSpace;

pub fn parseIndent(state: *BlockParserState) void {
    if (state.i < state.src.len and isSpace(state.src[state.i])) {
        const start = state.i;
        while (state.i < state.src.len) {
            const char = state.src[state.i];
            if (char == ' ') {
                state.indent += 1;
            } else if (char == '\t') {
                state.indent += 4 - @mod(state.indent, 4);
            } else if (char == '\n') {
                state.hasBlankLine = true;
                break;
            } else if (char == '\r') {
                state.hasBlankLine = true;
                if (state.i + 1 < state.src.len and state.src[state.i + 1] == '\n') {
                    state.i += 1;
                }
                break;
            } else {
                break;
            }
            state.i += 1;
        }
        state.spaces = state.src[start..state.i];
    }
}
